properties/extras/uid.py

```diff
@@ -0,0 +1,208 @@
+          in the input dictionary.
+        * Any other keyword arguments will be passed through to the Property
+          deserializers.
+
+        .. note::
+
+            UidModel instances are constructed with no input arguments
+            (ie :code:`cls()` is called). This means deserialization will
+            fail if the init method has been overridden to require
+            input parameters.
+        """
+        registry = kwargs.pop('registry', None)
+        if registry is None:
+            if not isinstance(value, dict):
+                raise ValueError('HasUID must deserialize from dictionary')
+            registry = value.copy()
+            uid = kwargs.get('uid', registry.get('__uid__'))
+        else:
+            uid = value
+        if uid in cls._INSTANCES and uid not in registry:
+            return cls._INSTANCES[uid]
+        elif uid in cls._INSTANCES:
+            raise ValueError('Uid already used: {}'.format(uid))
+        elif uid not in registry:
+            raise ValueError('Invalid uid: {}'.format(uid))
+        value = registry[uid]
+        if not isinstance(value, HasUID):
+            new_inst = cls()
+            registry.update({uid: new_inst})
+            super(HasUID, cls).deserialize(
+                value=value,
+                trusted=trusted,
+                strict=strict,
+                registry=registry,
+                _instance=new_inst,
+                **kwargs
+            )
+        cls._INSTANCES[uid] = registry[uid]
+        return registry[uid]
+
+
+class Pointer(base.Instance):
+    """Property for HasUID instances where string UID pointer may be used
+
+    **Available keywords** (in addition to those inherited from
+    :ref:`Instance <instance>`):
+
+    * **enforce_uid** - Require Pointer strings to resolve into instances.
+      If False, the default, the Pointer property may be set to an arbitrary
+      string; if True, the string must correspond to an existing UID.
+    """
+
+    @property
+    def instance_class(self):
+        """Allowed class for the Pointer property
+
+        Must be a subclass of HasUID
+        """
+        return self._instance_class
+
+    @instance_class.setter
+    def instance_class(self, value):
+        if not isinstance(value, CLASS_TYPES) or not issubclass(value, HasUID):
+            raise TypeError('instance_class must be a subclass of HasUID')
+        self._instance_class = value
+
+    @property
+    def enforce_uid(self):
+        """Require Pointer strings to resolve into instances
+
+        If False, the default, the Pointer may be set to an arbitrary string;
+        if True, the string must correspond to an existing UID.
+        """
+        return getattr(self, '_enforce_uid', False)
+
+    @enforce_uid.setter
+    def enforce_uid(self, value):
+        if not isinstance(value, bool):
+            raise TypeError('enforce_uid must be a boolean')
+        self._enforce_uid = value
```

Comment:
This doesn't feel very Pythonic to me. Why not just:

```python
self._enforce_uid = bool(value)
```

Reply:
Cool, fair enough. This comes up all over the library, actually. e.g. https://github.com/seequent/properties/blob/v0.5.0b1/properties/basic.py#L627 and I agree it's not very pythonic.

My only reasoning for this is that properties not only provides pretty strict type-checking and validation for users of the libraries, but also for developers building HasProperties classes. Especially in the HasProperties metaclass, there is some fairly strict micromanaging. Requiring `bool` encourages building the classes a specific way and gives a nice error otherwise.

Despite that reasoning, I think you are right, and remaining Pythonic should rule. Developers should be reading the docs, etc, and added flexibility but slightly-harder-to-follow errors is the right choice.

Reply:
At least for `bool`, maybe?
If we were only looking at Py3k we could use ABCs in theory.